### PR TITLE
Select: jsonpath support

### LINKED ISSUE #12: [2pt] Support JSON path
Even Tarantool 1.10 supports value extraction via jsonpath. So we could patch our filter generator and add jsonpath's support of jsonpath conditions.

### PR DESCRIPTION
**Pull request by mRrvz**
Added jsonpath support for select. Part of #12

### CODE REVIEW

File: crud/common/utils.lua [hunk 1]
```diff
@@ -5,6 +5,11 @@ local vshard = require('vshard')
 local schema = require('crud.common.schema')
 local dev_checks = require('crud.common.dev_checks')
 
+local keydef
+if _TARANTOOL >= "2" then
```
Comment (dokshina) [inline crud/common/utils.lua:9]: Take a look at `enabled_tarantool_features `(https://github.com/tarantool/crud/blob/master/crud/common/utils.lua#L169).

File: crud/common/utils.lua [hunk 2]
```diff
@@ -488,4 +493,35 @@ function utils.flatten_obj_reload(space_name, obj)
     return schema.wrap_func_reload(flatten_obj, space_name, obj)
 end
 
+function utils.format_jsonpath(operand, field_names, field_ids)
+    local ind = operand:find("]")
+    if ind == nil or keydef == nil then -- not jsonpath OR tarantool < 2
```
Comment (dokshina) [inline crud/common/utils.lua:498]: What does `ind` mean?
Are you sure that passing JSON path when Tarantool doesn't support it isn't an error?

File: crud/common/utils.lua [hunk 3]
```diff
@@ -488,4 +493,35 @@ function utils.flatten_obj_reload(space_name, obj)
...
+        return {
+            operand = operand,
+        }
+    end
+
+    local top_field = operand:sub(0, ind)
+    local ok, new_keydef = pcall(keydef.new, {{type = 'string', fieldno = 1, path = top_field}})
```
Comment (dokshina) [inline crud/common/utils.lua:505]: `keydef` is a module, but `new_keydef` is a keydef object.

Use `key_def_lib = require('tuple.keydef')` like in `crud/select/merger.lua`.

File: crud/common/utils.lua [hunk 4]
```diff
@@ -488,4 +493,35 @@ function utils.flatten_obj_reload(space_name, obj)
...
+        return nil, ParseOperationsError:new("Invalid jsonpath format")
+    end
+
+    local ok, field_name = pcall(new_keydef.extract_key, new_keydef, box.tuple.new{field_names})
+    if not ok then
+        return nil, ParseOperationsError:new("Space doesn't contains field %s", top_field)
+    end
+
+    local jsonpath = nil
+    local nested_field = operand:sub(ind + 1, #operand)
```
Comment (dokshina) [inline crud/common/utils.lua:516]: Check out https://github.com/olegrok/tarantool-jsonpath.

File: crud/select/filters.lua [hunk 5]
```diff
@@ -100,22 +100,26 @@ local function parse(space, conditions, opts)
...
+                local fieldno = fieldnos_by_names[condition.operand]
+                fieldnos = {fieldno}
+                local field_format = space_format[fieldno]
                 fields_types = {field_format.type}
                 local is_nullable = field_format.is_nullable == true
                 values_opts = {
                     {is_nullable = is_nullable, collation = nil},
                 }
+
+                --jsonpath = {condition.jsonpath}
```
Comment (dokshina) [inline crud/select/filters.lua:122]: ?

File: test/integration/simple_operations_test.lua [hunk 6]
```diff
@@ -408,7 +408,7 @@ pgroup:add('test_intermediate_nullable_fields_update', function(g)
 
     helpers.call_on_servers(g.cluster, {'s1-master', 's2-master'}, function(server)
         for i = 1, 12 do
-            server.net_box:call('add_extra_field', {'extra_' .. tostring(i)})
+            server.net_box:call('add_extra_field', {'extra_' .. i})
```
Comment (dokshina) [inline test/integration/simple_operations_test.lua:411]: Why?

File: crud/common/utils.lua [hunk 7]
```diff
@@ -181,6 +181,9 @@ local function determine_enabled_features()
 
     -- since Tarantool 2.4
     enabled_tarantool_features.uuids = major >= 2 and (minor > 4 or minor == 4 and patch >= 1)
+
+    -- since Tarantool 2.0
+    enabled_tarantool_features.jsonpaths = major >= 2
```
Comment (olegrok) [inline crud/common/utils.lua:186]: Since 2.6.3/2.7.2/2.8.1 because of https://github.com/tarantool/tarantool/pull/5497

Comment (olegrok) [reply crud/common/utils.lua:186]: Here we should divide "jsonpath filters" that are available since 1.10 and "json path indexes" that was introduced in 2.1 but you can't it use in correct way because of a bug specified above.

File: crud/select/filters.lua [hunk 8]
```diff
@@ -408,8 +437,9 @@ local function gen_filter_code(filter_conditions)
     table.insert(filter_code_parts, '')
 
     for i, cond in ipairs(filter_conditions) do
-        local args_fieldnos = { unpack(cond.fieldnos, 1, #cond.values) }
-        local func_args_code = function_args_by_fieldnos(args_fieldnos)
+        local args_fields = { unpack(cond.fields, 1, #cond.values) }
+        --local args_fields = get_args_fields(cond)
```
Comment (olegrok) [inline crud/select/filters.lua:441]: Drop redundant comment

File: test/entrypoint/srv_select.lua [hunk 9]
```diff
@@ -126,6 +126,40 @@ package.preload['customers-storage'] = function()
...
+            })
+
+            developers_space:create_index('full_name', {
+                parts = {
+                    {field = 'name'},
+                    {field = 'last_name'},
+                },
+                unique = false,
+                if_not_exists = true,
+            })
```
Comment (olegrok) [inline test/entrypoint/srv_select.lua:162]: Where is at least one jsonpath index?

File: test/unit/parse_conditions_test.lua [hunk 10]
```diff
@@ -101,3 +102,27 @@ g.test_parse_errors = function()
         'condition[2] should be string, got "number" (condition 2)'
     )
 end
+
+g.test_jsonpath_parse = function()
+    t.skip_if(not crud_utils.tarantool_supports_jsonpaths(), "Jsonpath is not supported on Tarantool < 2")
```
Comment (olegrok) [inline test/unit/parse_conditions_test.lua:107]: It's filters it's available since 1.10. But you don't have indexes with "path" - see https://github.com/tarantool/doc/issues/708

Comment (olegrok) [reply test/unit/parse_conditions_test.lua:107]: Here is a doc ticket about json-path access inside tuple - https://github.com/tarantool/doc/issues/501

File: test/entrypoint/srv_select.lua (hunk 9 above)
Comment (olegrok) [reply test/entrypoint/srv_select.lua:162]: I expect to see following tests:
  - primary key is jsonpath
  - secondary key is jsonpath
  - pagination for indexes above
  - sharding key calculation for such indexes (aka bucket optimization)

File: crud/common/utils.lua [hunk 11]
```diff
@@ -181,6 +181,13 @@ local function determine_enabled_features()
 
     -- since Tarantool 2.4
     enabled_tarantool_features.uuids = major >= 2 and (minor > 4 or minor == 4 and patch >= 1)
+
+    -- since Tarantool 1.10
+    enabled_tarantool_features.jsonpath_filters = major >= 2 or (major >= 1 and minor >= 10)
+
+    -- since Tarantool 2.6.3 / 2.7.2 / 2.8.1
+    enabled_tarantool_features.jsonpath_indexes = major >= 2 and ((minor >= 6 and patch >= 3)
+        or (minor >= 7 and patch >= 2) or (minor >= 8 and patch >= 1) or minor >= 9)
```
Comment (olegrok) [inline crud/common/utils.lua:190]: 3.0.0 > 2.6.3

```
tarantool> major = 3
---
...

tarantool> minor = 0
---
...

tarantool> patch = 0
---
...

tarantool> major >= 2 and ((minor >= 6 and patch >= 3) or (minor >= 7 and patch >= 2) or (minor >= 8 and patch >= 1) or minor >= 9)
---
- false
...
```

File: crud/common/utils.lua [hunk 12]
```diff
@@ -181,6 +181,13 @@ local function determine_enabled_features()
 
     -- since Tarantool 2.4
     enabled_tarantool_features.uuids = major >= 2 and (minor > 4 or minor == 4 and patch >= 1)
+
+    -- since Tarantool 1.10
+    enabled_tarantool_features.jsonpath_filters = major >= 2 or (major >= 1 and minor >= 10)
```
Comment (olegrok) [inline crud/common/utils.lua:186]: It doesn't make sense since crud supports only 1.10+.

You couldn't run crud e.g. on Tarantool 1.9 even before this patch.

File: test/integration/select_test.lua [hunk 13]
```diff
@@ -1057,6 +1058,158 @@ pgroup:add('test_cut_selected_rows', function(g)
...
+            id = {dev_id = 4},
+            manufacturer = "General Motors",
+        },
+    }
+    t.assert_equals(objects, expected_objects)
+
+    -- Secondary jsonpath index
+    local result, err = g.cluster.main_server.net_box:call('crud.select',
+        {'cars', {{'==', 'data_index', 'Yellow'}}})
+    t.assert_equals(err, nil)
```
Comment (olegrok) [inline test/integration/select_test.lua:1174]: I don't see any test of pagination for such indexes.

File: crud/common/utils.lua [hunk 14]
```diff
@@ -148,6 +152,65 @@ function utils.extract_key(tuple, key_parts)
...
+
+    if keydef_cache[key] ~= nil then
+        return keydef_cache[key]
+    end
+
+    local key_def = keydef_lib.new(index_parts)
+    keydef_cache[key] = key_def
+
+    return key_def
+end
```
Comment (olegrok) [inline crud/common/utils.lua:169]: It starts to look stranger and stranger. Is msgpack.encode with utils.get_index_paths really faster than key_def.new?

File: crud/compare/comparators.lua [hunk 15]
```diff
@@ -152,9 +153,13 @@ function comparators.gen_tuples_comparator(cmp_operator, key_parts, field_names,
...
-    return function(lhs, rhs)
-        local lhs_key = utils.extract_key(lhs, updated_key_parts)
-        local rhs_key = utils.extract_key(rhs, updated_key_parts)
+    return function(lhs, rhs, skip_extract)
+        if skip_extract == true then
+            return keys_comparator(lhs, rhs)
+        end
+
+        local lhs_key = utils.extract_jsonpath_keys(lhs, updated_key_parts)
+        local rhs_key = utils.extract_jsonpath_keys(rhs, updated_key_parts)
```
Comment (olegrok) [inline crud/compare/comparators.lua:162]: Do you really need it? This code should be executed only at versions where merger is not supported. But such versions usually don't support json-path indexes yet.

File: crud/select/executor.lua [hunk 16]
```diff
@@ -25,7 +25,7 @@ local function scroll_to_after_tuple(gen, space, scan_index, tarantool_iter, aft
             return nil
         end
 
-        if scroll_comparator(tuple, after_tuple) then
+        if scroll_comparator(tuple:totable(), after_tuple) then
```
Comment (olegrok) [inline crud/select/executor.lua:28]: Why do you need this change?

File: crud/select/filters.lua [hunk 17]
```diff
@@ -42,11 +40,15 @@ local function is_early_exit_possible(index, tarantool_iter, condition)
     return false
 end
 
-local function get_index_fieldnos(index)
+local function get_index_fieldnos(index, space_format)
```
Comment (olegrok) [inline crud/select/filters.lua:43]: Is it possible to move this changes into separate patch?

**Review by olegrok — APPROVED** [review]
Thanks for your patch. LGTM.

Please change description of your PR - it doesn't "closes #12", it's only part of it.

**Review by dokshina — APPROVED** [review]
Thanks for your patch!

I am sure that we should update select examples in the doc. Of course, in a separated patch.
